networkx/algorithms/approximation/steinertree.py

```diff
@@ -114,18 +114,29 @@ def _kou_steiner_tree(G, terminal_nodes, weight):
 
 def _remove_nonterminal_leaves(G, terminals):
     terminals_set = set(terminals)
-    degree_one_nodes = {n for n in G if G.degree(n) == 1}
-    nonterminal_leaves = degree_one_nodes - terminals_set
+    self_looped_nodes = {n for n in G if n in G[n] and len(set(G[n])) == 2}
+    degree_1_not_looped = {n for n in G if (G.degree(n) == 1) and (n not in G[n])}
```

Comment:
Hmmm yes, I like what you're thinking, only, I don't want to have to iterate through the whole graph on each iteration, because if you have a large graph (like what I hope to use this for in the future) this will start to get expensive quick. I don't want to pay O(k*n), where k is undetermined, for something that's providing a very small improvement on an approximation. 

My idea is to leverage the fact that we know that if we remove a leaf, that we only have to check that leaf's neighbors for if they should be removed as well, to inform us to iterate only through those. That's what the `possible_nonterminal_leaves` is supposed to be doing. 

I totally understand what you're saying though saying, `self_looped_nodes` isn't explaining what the cases are, and my approach was a little over-complicated. I've simplified the leaf calculating step significantly, inspired by your suggestion, while still only iterating through the possibly newly created leaves. 

Thanks very much for your comments, you've given me some good ideas.